datastore/datastore.go

```diff
@@ -0,0 +1,260 @@
+
+// Options are options for the Hydra datastore
+type Options struct {
+	// total number of find provider queries we should queue
+	FindProvidersQueueSize int
+	// number of providers to find when a provider record does not exist in the store
+	FindProvidersCount int
+	// number of find provider requests we will concurrently process
+	FindProvidersConcurrency int
+	// maximum time a find providers call is allowed to take
+	FindProvidersTimeout time.Duration
+	// period after a find failure that another find request for the same key will be discarded
+	FindProvidersFailureBackoff time.Duration
+}
+
+// option defaults
+const (
+	findProvidersQueueSize   = 1000
+	findProvidersCount       = 1
+	findProvidersConcurrency = 1
+	findProvidersTimeout     = time.Second * 10
+)
+
+type findResult struct {
+	key      datastore.Key
+	status   string
+	duration time.Duration
+	err      error
+}
+
+// NewDatastore creates a new datastore that adds hooks to perform hydra things
+func NewDatastore(ctx context.Context, path string, getRouting GetRoutingFunc, opts Options) (datastore.Batching, error) {
+	opts = setOptionDefaults(opts)
+	ds, err := levelds.NewDatastore(path, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create datastore: %w", err)
+	}
+
+	return hook.NewBatching(ds, hopts.OnAfterQuery(newOnAfterQueryHook(ctx, getRouting, opts))), nil
+}
+
+func setOptionDefaults(opts Options) Options {
+	if opts.FindProvidersConcurrency == 0 {
+		opts.FindProvidersConcurrency = findProvidersConcurrency
+	}
+	if opts.FindProvidersCount == 0 {
+		opts.FindProvidersCount = findProvidersCount
+	}
+	if opts.FindProvidersQueueSize == 0 {
+		opts.FindProvidersQueueSize = findProvidersQueueSize
+	}
+	if opts.FindProvidersTimeout == 0 {
+		opts.FindProvidersTimeout = findProvidersTimeout
+	}
+	if opts.FindProvidersFailureBackoff == 0 {
+		opts.FindProvidersFailureBackoff = time.Minute
+	}
+	return opts
+}
+
+func newOnAfterQueryHook(ctx context.Context, getRouting GetRoutingFunc, opts Options) hopts.OnAfterQueryFunc {
+	findC := make(chan datastore.Key, opts.FindProvidersQueueSize)
+	foundC := make(chan findResult)
+
+	pending := make(map[string]bool)
+	var pendingLock sync.Mutex
+
+	failed := timecache.NewTimeCache(opts.FindProvidersFailureBackoff)
+	var failedLock sync.Mutex
+
+	for i := 0; i < opts.FindProvidersConcurrency; i++ {
+		go findProviders(ctx, findC, foundC, getRouting, opts)
+	}
+
+	go func() {
+		for {
+			select {
+			case r := <-foundC:
+				pendingLock.Lock()
+				delete(pending, r.key.String())
+				stats.Record(ctx, metrics.FindProvsQueueSize.M(-1))
+				pendingLock.Unlock()
+
+				if r.err != nil {
+					fmt.Println(r.err)
+					continue
+				}
+
+				recordFindProvsComplete(ctx, r.status, metrics.FindProvsDuration.M(float64(r.duration)))
+
+				// if we failed to find this key, add to the failure cache so we don't try again for a bit
+				if r.status == "failed" {
+					failedLock.Lock()
```

Comment:
Adding stuff to the timecache requires a lock? What is the concern behind this lock?

Reply:
I don’t want to check for an item at the same time as one is being added. If the check comes back false, then the item is added we’ll try to add the same item again and it’ll panic.

Could probably be a RWLock but I don’t think this is a bottleneck.